core/domain/state_domain.py

```diff
@@ -916,8 +938,946 @@ def get_all_outcomes(self) -> List[Outcome]:
+        allow_non_zero_integ_part = (
+            self.customization_args['allowNonzeroIntegerPart'].value)
+        allow_imp_frac = self.customization_args['allowImproperFraction'].value
+        require_simple_form = (
+            self.customization_args['requireSimplestForm'].value)
+
+        for ans_group_index, answer_group in enumerate(self.answer_groups):
+            for rule_spec_index, rule_spec in enumerate(
+                answer_group.rule_specs
+            ):
+                # Rule should not be duplicate.
+                if rule_spec.to_dict() in rule_spec_till_now:
+                    raise utils.ValidationError(
+                        f'The rule \'{rule_spec_index}\' of answer group '
+                        f'\'{ans_group_index}\' of FractionInput '
+                        f'interaction is already present.'
+                    )
+                rule_spec_till_now.append(rule_spec.to_dict())
+
+                if rule_spec.rule_type not in inputs_without_fractions:
+                    num = rule_spec.inputs['f']['numerator']
+                    den = rule_spec.inputs['f']['denominator']
+                    whole = rule_spec.inputs['f']['wholeNumber']
+
+                    # Solution should be in simplest form if the `simplest form`
+                    # setting is turned on.
+                    if require_simple_form:
+                        d = math.gcd(num, den)
+                        val_num = num // d
+                        val_den = den // d
+                        if val_num != num and val_den != den:
+                            raise utils.ValidationError(
+                                f'The rule \'{rule_spec_index}\' of '
+                                f'answer group \'{ans_group_index}\' do '
+                                f'not have value in simple form '
+                                f'in FractionInput interaction.'
+                            )
+
+                    # Solution should be in proper form if the `proper form`
+                    # setting is turned on.
+                    if not allow_imp_frac:
+                        if whole == 0:
+                            if den <= num:
+                                raise utils.ValidationError(
+                                    f'The rule \'{rule_spec_index}\' of '
+                                    f'answer group \'{ans_group_index}\' do '
+                                    f'not have value in proper fraction '
+                                    f'in FractionInput interaction.'
+                                )
+                        else:
+                            raise utils.ValidationError(
+                                f'The rule \'{rule_spec_index}\' of '
+                                f'answer group \'{ans_group_index}\' do '
+                                f'not have value in proper fraction '
+                                f'in FractionInput interaction.'
+                            )
```

Comment:
This would be IMO more readable
```suggestion
                    if not allow_imp_frac and whole != 0:
                        raise utils.ValidationError(
                            f'The rule \'{rule_spec_index}\' of '
                            f'answer group \'{ans_group_index}\' do '
                            f'not have value in proper fraction '
                            f'in FractionInput interaction.'
                        )
                        
                        if not allow_imp_frac and den <= num:
                            raise utils.ValidationError(
                                f'The rule \'{rule_spec_index}\' of answer 
                                f'group \'{ans_group_index}\' do '
                                f'not have value in proper fraction '
                                f'in FractionInput interaction.'
                            )
```

you might also want to reorganize the strings as we moved a bit more to the left

Reply:
Done